### PR TITLE
WIP: Pixit parser

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
chip-tool does not have a handler for PIXIT values

#### Change overview
Adds PIXIT handling to chip tool
- parser to transform pixit file (same format as PICS file) into proper types
- reader to read pixit values from file
- adds pixit storage to TestCommand

NOTE: darwin changes not yet done, opening PR to get some early feedback

#### Testing
- TypeTransformer includes unit tests
- Reader includes unit tests
- chip-tool tested using yaml test
- read from file in chip tool tested manually against yaml test (chip-tool tests TestPixit --PIXIT <file>)

### CODE REVIEW

File: src/app/zap-templates/templates/app/helper.js [hunk 1]
```diff
@@ -849,6 +849,67 @@ async function if_is_non_zero_default(value, options)
...
+      return "chip::CharSpan(" + pixitInC + ".c_str(), " + pixitInC + ".length())";
+    } else {
+      return pixitInC + ".c_str()";
+    }
+  } else {
+    return pixitInC;
+  }
+}
+
+async function maybeSubPixit(val, type, typedLiteral, useSpans)
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/helper.js:900]: Fwiw, if this function signature were:
```
 async function maybeSubPixit(val, type, otpions)
```
then the callsites could look like:
```
{{maybeSubPixit value ../type}}
{{maybeSubPixit value ../type typedLiteral=true}}
```
and the check in this function would be `if (options.hash.typedLiteral)`.

File: src/app/zap-templates/templates/app/helper.js [hunk 2]
```diff
@@ -849,6 +849,67 @@ async function if_is_non_zero_default(value, options)
...
+{
+  if (isPixit(val)) {
+    return await formatPixit.call(this, val, type, useSpans);
+  } else if (typedLiteral) {
+    return await asTypedLiteral.call(this, val, type);
+  } else if (StringHelper.isOctetString(type) || StringHelper.isCharString(type)) {
+    return "\"" + val + "\"";
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/helper.js:907]:
```suggestion
    return `"${val}"`;
```
is a lot more readable.

File: src/app/zap-templates/templates/app/helper.js [hunk 3]
```diff
@@ -849,6 +849,67 @@ async function if_is_non_zero_default(value, options)
...
+  return "mPixit[\"" + val + "\"].Get<" + basicType + ">()"
+}
+
+async function formatPixit(val, type, useSpans)
+{
+  const pixitInC = await PixitInC.call(this, val, type);
+  if (StringHelper.isOctetString(type)) {
+    if (useSpans) {
+      // This will currently do weird things if the pixit is a hex string. So...don't do that.
+      return "chip::ByteSpan(chip::Uint8::from_const_char(" + pixitInC + ".c_str()), " + pixitInC + ".length())";
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/helper.js:884]:
```suggestion
      return `chip::ByteSpan(chip::Uint8::from_const_char(${pixitInC}.c_str()), ${pixitInC}.length())`;
```

File: src/app/zap-templates/templates/app/helper.js [hunk 4]
```diff
@@ -849,6 +849,67 @@ async function if_is_non_zero_default(value, options)
...
+    } else {
+      return pixitInC + ".c_str()";
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/helper.js:886]: Or could do 
```
`${pixitInC}.c_str()`
```

File: src/app/zap-templates/templates/app/helper.js [hunk 5]
```diff
@@ -849,6 +849,67 @@ async function if_is_non_zero_default(value, options)
...
+    }
+  } else if (StringHelper.isCharString(type)) {
+    if (useSpans) {
+      // This should be the UTF-8 encode length, but it's just the length. So, don't use weird chars.
+      return "chip::CharSpan(" + pixitInC + ".c_str(), " + pixitInC + ".length())";
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/helper.js:891]:
```suggestion
      return `chip::CharSpan(${pixitInC}.c_str(), ${pixitInC}.length())`;
```

**Review by bzbarsky-apple — APPROVED** [review]
I didn't review the actual parser very carefully; mostly focused on the ZAP bits.

**Comment by cecille** [comment]
Thanks @bzbarsky-apple - if you're ok with the general approach here, I can take a stab at getting the darwin parts working. Just didn't want to start in on that before there was a general OK since it'll take me a bit longer to get set up and test.

**Comment by cecille** [comment]
Marking this draft so it doesn't get accidentally added. Not that it would - darwin tests won't pass. Still, better safe than sorry

**Comment by bzbarsky-apple** [comment]
@cecille The general approach seems pretty sane to me, I think.

**Comment by cecille** [comment]
ahoy, stalebot, please do not close this - I was on vacation.

**Comment by cecille** [comment]
Going to abandon this change for now. In theory, I like the approach of having a PIXIT file, but now that SVE2 is done and we have some breathing room, it makes sense to step back and consider this entire system more holistically. This PR has shows that there are some gaps in the darwin feature set, and I'd rather not make this stack more fragile than it actually is.

**Pull request closed without merging**